### PR TITLE
Change dim `axis_label` resize logic to set width using only displayed labels width

### LINKED ISSUE #5521: Slider widget resizing when changing axes labels
## 🐛 Bug

Sliders are resizing when adjusted with `ViewerModel.dims.axis_labels` (thanks @andy-sweet for pointing out the source of the bug).

Here is an example of what it looks like.

Starting off normally:
image

When typing the label the slider resizes:
image

_Originally posted by @kephale in https://github.com/andy-sweet/napari-metadata/issues/8#issuecomment-1410515298_

@andy-sweet made [some comments](https://github.com/andy-sweet/napari-metadata/pull/8#issuecomment-1410749226) about this bug:

> Good catch. That's a core napari bug, which you can reproduce by setting `ViewerModel.dims.axis_labels`. I suspect it's due to a slider widget existing for each axis/dimension and being laid out, but also being hidden. There's nothing we can do about it here, but might be worth creating an issue on napari/napari.
            
## To Reproduce

1. Open an image that has a slider, like cells 3d
2. Set `ViewerModel.dims.axis_labels` 

## Expected behavior

The slider shouldn't resize

### PR DESCRIPTION
**Pull request by dalthviz**
# Description

Old behavior:

GIF showing the old behavior of the viewer slider widgets axis label resize. The labels take the width of the longest axis label element even if it is not being displayed. The space for the axis labels is the same even if the current axis label text is shorter than the space available

New behavior:

GIF showing the new behavior introduced with the PR for the viewer slider widgets axis label resize. The labels take the width of the longest axis label element that is being displayed. The alignment between multiple slider widgets is preserved but changing the axis being displayed can make the axis labels to take more or less space

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# References

closes #5521 

# How has this been tested?

- [x] all tests pass with my change
- [x] I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: napari/_qt/widgets/qt_dims.py
Comment (andy-sweet) [inline napari/_qt/widgets/qt_dims.py]: Minor: the list of indices is only used once and accessing the widgets should be cheap, so just combine the statements to make this more concise.

```suggestion
        displayed_labels = [
            self.slider_widgets[idx].axis_label
            for idx, displayed in enumerate(self._displayed_sliders)
            if displayed
        ]
        if displayed_labels:
            fm = QFontMetrics(QFont("", 0))
            labels = self.findChildren(QLineEdit, 'axis_label')
```

File: napari/_qt/widgets/qt_dims.py
Comment (andy-sweet) [inline napari/_qt/widgets/qt_dims.py:144]: Question: why use `QFontMetrics(QFont("", 0))` instead of `self.fontMetrics()`? It's unclear to me where the family and point size of the font come from in the former, whereas the latter should at least pick the current family/point size from the top level widget here.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Thanks for the fix here! A nice, small improvement. The only change I requested is minor.

On my machine, this fix almost worked, but some text is also lost on the left edge (the full text is `purpleee...` as in the original bug) when the label gets really long. I suspect this is rare in practice, so is not worth too much thought, but I wonder if it could be related to our use of `QFontMetrics` or the magic numbers (0.2, 10) in the width calculation (neither of which are new to this PR).

Screenshot 2023-03-20 at 8 37 53 AM

File: napari/_qt/widgets/qt_dims.py
Comment (dalthviz) [reply napari/_qt/widgets/qt_dims.py:144]: I'm not totally sure 🤔 I just took the way this was being done before to get the `QFontMetrics` instance, but that's a good point. Should I try to use `self.fontMetrics()`?

**Review by andy-sweet — APPROVED** [review]
Looks good to me now. I'll wait about 48 hours before merging this in case others have different opinions or want to fix some of the related issues in this PR.

File: napari/_qt/widgets/qt_dims.py
Comment (andy-sweet) [reply napari/_qt/widgets/qt_dims.py:144]: For this PR, I don't need you to since that usage was around anyway and is independent of the bug being fixed here, so I've approved.

However, if you're interested into digging into this more in this PR or elsewhere, I won't say no either. Also happy to help debug this by executing stuff on my machine and environments in case there's some dependence there.

**Comment by dalthviz** [comment]
Thanks for the review @andy-sweet ! I will give it a try and check a fix for the related issues mentioned :+1:

**Comment by dalthviz** [comment]
Quick update!

I used `self.fontMetrics` and tested doing some other changes for the extra width addition (the `+10` that was around) and also adding some eliding to the text that takes more than 20% of the whole slider widget size (the `* 0.2` that is being done to determine a max width) and ended up with something like this:

```python
    def _resize_axis_labels(self):
        """When any of the labels get updated, this method updates all label
        widths to a minimum size. This allows the full label to be
        visible at all times, with minimal space, without setting stretch on
        the layout.
        """
        displayed_labels = [
            self.slider_widgets[idx].axis_label
            for idx, displayed in enumerate(self._displayed_sliders)
            if displayed
        ]
        if displayed_labels:
            fm = self.fontMetrics()
            labels = self.findChildren(QLineEdit, 'axis_label')
            # set maximum width to no more than 20% of slider width
            maxwidth = int(self.slider_widgets[0].width() * 0.2)
            # set new base width to the width of the longest label being displayed
            newwidth = max(
                [
                    int(fm.boundingRect(dlab.text()).width())
                    for dlab in displayed_labels
                ]
            )

            for labl in labels:
                # some extra margin for the label text
                # here the average width of a character is used as base measure
                # to add some extra width. We double it so we take into account
                # a space before and after the text
                margin_width = int(fm.averageCharWidth() * 2)
                base_labl_width = min([newwidth, maxwidth])
                labl_width = base_labl_width + margin_width
                if base_labl_width == maxwidth:
                    elide_text = fm.elidedText(
                        labl.text(), Qt.ElideRight, labl_width
                    )
                    labl.setText(elide_text)
                    # give more margin space due to the `...` elide symbol so
                    # we use the average character width but three times
                    margin_width = int(fm.averageCharWidth() * 3)
                    labl.setFixedWidth(labl_width + margin_width)
                else:
                    labl.setFixedWidth(labl_width)

```
Also I had to add some logic to be able to detect a slider resize event to update the `axis_label` width with something like this in the `QtDimSliderWidget` class:

```python
    def resizeEvent(self, event):
        self.resized.emit()
        super().resizeEvent(event)
```

A preview:

axis_label_elide

Still there is some work I will need to do in order to be able to restore the text if eliding is not necessary so that makes me think that probably a subclass of `QLineEdit` should be created to handle the `axis_label` with eliding capabilities for the case where the text takes more than 20% of the whole slider size (if adding eliding capabilities to the axis label sounds like a good idea).

Following that I think it makes more sense to merge this as it is and then, in a follow up PR, I could implemented the fix/enhancement described above to prevent cutting the axis label text :)

**Comment by andy-sweet** [comment]
> Following that I think it makes more sense to merge this as it is and then, in a follow up PR, I could implemented the fix/enhancement described above to prevent cutting the axis label text :)

Agreed.

Could you expand the text a little in the PR description to give a short summary of what's happening in the gifs? Or alternatively, provide alt-text to the same effect. We're using the PR description as the merge commit message now, so it's quite common to see the text only version of that.

Then I'll merge this tomorrow unless I hear otherwise.

**Comment by dalthviz** [comment]
Updated the alt-text for the GIFs to:

* Old behavior:
```
GIF showing the old behavior of the viewer slider widgets axis label resize. The labels take the width of the longest axis label element even if it is not being displayed. The space for the axis labels is the same even if the current axis label text is shorter than the space available
```
* New behavior:
```
GIF showing the new behavior introduced with the PR for the viewer slider widgets axis label resize. The labels take the width of the longest axis label element that is being displayed. The alignment between multiple slider widgets is preserved but changing the axis being displayed can make the axis labels to take more or less space
```

**Pull request merged**
